### ISSUE TITLE
Exec Terminal in UI not working for some deployments

### ISSUE THREAD

**Issue by siccous** [hashicorp/waypoint#2948]
**Describe the bug**
For some of my apps in Waypoint the Exec Terminal is not able to connect. They are deployed throught the Nomad jobspec plugin.

**Steps to Reproduce**
Simply try to use the Exec Terminal and it hangs with:
![image](https://user-images.githubusercontent.com/7812885/151153481-dfdb2e35-f4af-45a6-82c7-ee8d62320db1.png)

Stderr from Waypoint server shows this:
![image](https://user-images.githubusercontent.com/7812885/151153642-8f95b79a-aec7-4c98-8691-c0c3c3923dae.png)

**Waypoint Platform Versions**
Additional version and platform information to help triage the issue if
applicable:

* Waypoint CLI Version: 0.7.1
* Waypoint Server Platform and Version: (like `docker`, `nomad`, `kubernetes`): 0.7.1 - Nomad jobspec
* Waypoint Plugin: (like `aws/ecs`, `pack`, `azure`): pack

**Additional context**
Right now we have 4 apps deployed through Waypoint and it doesn't work for 2 of them.

**Comment by krantzinator** [comment]
Are the 2 deployments that this does work for also through nomad job spec?
Can you share the waypoint.hcls and nomad job file? The problem may be that the entrypoint is not being embedded in the built image, and/or that the built image with the entrypoint is not being shared with the jobspec plugin.

**Comment by siccous** [comment]
Yep, all of them works the same, the only difference is only the app itself which is deployed.

I've redacted the waypoint.hcl and nomad job a little bit (replaced all sensitiver stuff with ...)

[tim-job.nomad.tpl.txt](https://github.com/hashicorp/waypoint/files/7949497/tim-job.nomad.tpl.txt)
[waypoint.hcl.txt](https://github.com/hashicorp/waypoint/files/7949498/waypoint.hcl.txt)

**Comment by siccous** [comment]
Heh, just tried to show it to my colleague and suddenly the Exec terminal stopped working for all of our deploments.

So, it's not 2/4 now but 0/4. Only difference are few new deployments of all of them :) No changes in waypoint.hcl or Nomad job template.

**Comment by krantzinator** [comment]
Can you share the files in gists so we don't have to download them? Thanks :)

**Comment by siccous** [comment]
https://gist.github.com/siccous/6804a07ead7b70232c49576976c69c51
https://gist.github.com/siccous/65b360bdc6ed0b614c2f8571ef2b91c9

**Comment by krantzinator** [comment]
Thank you! I'm going to try to recreate this, but quickly wanted to confirm that running `exec` on the CLI also gives you the same error as trying to access Exec in the UI?

**Comment by siccous** [comment]
Good catch, it works from CLI:

![image](https://user-images.githubusercontent.com/7812885/151433601-55557eac-35ba-484d-88fc-383c0d26f262.png)

**Comment by krantzinator** [comment]
OK I am able to replicate this just on a local nomad cluster. I have a hunch this is not plugin-specific, so I'll try to replicate on a local Docker setup as well.
Also in mine, it appears to break over time, as you have seen. I added a couple of apps to my hcl and eventually, after accessing `exec` a few times, saw one fail.

I _am_ seeing the failures in both the UI and the CLI though.

**Comment by krantzinator** [comment]
In the files you shared, I just see one app in the waypoint.hcl. Are you deploying this same app multiple times?
For the apps that you cannot exec into, can you access their deployment URLs? I'm wondering if there's a collision happening here in the configs.

**Comment by siccous** [comment]
Nope, we are deploying 4 different apps but all of them use very similar files. But every app has it's own waypoint.hcl and nomad job template.

I've disabled the deployment URL feauture few days ago on the server and never used it even before. But it didn't work before the switch off too.

**Comment by evanphx** [comment]
@siccous Could you provide us the output from looking at the fully installed job file that nomad is using? Specifically we're looking to see what environment variables the job got injected with.

**Comment by evanphx** [comment]
@siccous How did you setup nomad btw? There is an issue with the nomad setup where by default, it uses the specific allocation address for entrypoints to connect to (as there is no other by-default value). The cause of this is if the server restarts and has a new allocation, it can have a new IP/port and thusly the entrypoints can't connect back to the server.

So check the address that the nomad server is using and also what the ADVERTISE variables are in the job spec, it's possible they don't match.

**Comment by siccous** [comment]
> @siccous Could you provide us the output from looking at the fully installed job file that nomad is using? Specifically we're looking to see what environment variables the job got injected with.

Not exactly sure what do you need?

![image](https://user-images.githubusercontent.com/7812885/152229399-3c531ef2-f792-435f-bac1-226b1b5bfcb5.png)

> @siccous How did you setup nomad btw? There is an issue with the nomad setup where by default, it uses the specific allocation address for entrypoints to connect to (as there is no other by-default value). The cause of this is if the server restarts and has a new allocation, it can have a new IP/port and thusly the entrypoints can't connect back to the server.
> 
> So check the address that the nomad server is using and also what the ADVERTISE variables are in the job spec, it's possible they don't match.

I don't think there is issue with this, entrypoint can connect to Waypoint server (we are injecting env variables from Vault in waypoint.hcl and they are there)

**Comment by siccous** [comment]
> OK I am able to replicate this just on a local nomad cluster. I have a hunch this is not plugin-specific, so I'll try to replicate on a local Docker setup as well.

@krantzinator @evanphx anything I can do more to help you with this?
